lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const ComputedArtifact = require('./computed-artifact');
+const TracingProcessor = require('../../lib/traces/tracing-processor');
+
+const LONELY_TASK_FMP_DISTANCE = 5000;
+const LONELY_TASK_ENVELOPE_SIZE = 250;
+const LONELY_TASK_NEIGHBOR_DISTANCE = 1000;
+
+const MAX_QUIET_WINDOW_SIZE = 5000;
+const TRACE_BUSY_MSG = 'trace was busy the entire time';
+
+/**
+ * @fileoverview This artifact identifies the time the page is "first interactive" as defined below
+ * @see https://docs.google.com/document/d/1GGiI9-7KeY3TPqS3YT271upUVimo-XiL5mwWorDUD4c/edit#
+ *
+ * First Interactive marks the first moment when a website is minimally interactive:
+ *    > Enough (but maybe not all) UI components shown on the screen are interactive
+ *    > The page responds to user input in a reasonable time on average, but it’s ok if this
+ *      response is not always immediate.
+ *
+ * First Interactive is defined as the first period after FMP of N-seconds where there are no
+ * long tasks that are not "lonely".
+ *
+ *    > t = time in seconds since FMP
+ *    > N = 4 * e^(-0.045 * t) + 1
+ *    > a "lonely" task is an envelope of 250ms that contains a set of long tasks that have at least
+ *      1 second of padding before and after the envelope that contain no long tasks.
+ */
+class FirstInteractive extends ComputedArtifact {
+  get name() {
+    return 'FirstInteractive';
+  }
+
+  /**
+   * @param {number} t The time passed since FMP in miliseconds.
+   * @return {number}
+   */
+  static getRequiredWindowSizeInMs(t) {
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
+    return (4 * exponentiationComponent + 1) * 1000;
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number>}} longTasks
+   * @param {number} i
+   * @param {number} FMP
+   * @param {number} traceEnd
+   * @return {number} The last index of the lonely task envelope,
+   *    -1 if not a lonely task
+   */
+  static getLastLonelyTaskIndex(longTasks, i, FMP, traceEnd) {
```

Comment:
for just `getLastLonelyTaskIndex` or the whole thing? The trickier piece here that makes everything annoying is the complicated base case differentiation. I planned to fix that by treating the entire span before FMP as one giant long task which will make the base case the same as considering the first long task, but not sure if I should do in this PR. If it's too hard to grok right now I can do that switch now instead of a follow-up, wdyt?

FWIW, I totally agree it's not nearly as easy to follow, but at least a tiny bit that's because lonely tasks are way harder to explain that simple quiet windows too :)